shell/platform/linux/fl_engine.cc

```diff
@@ -34,6 +35,9 @@ static constexpr size_t kPlatformTaskRunnerIdentifier = 1;
 static constexpr int32_t kMousePointerDeviceId = 0;
 static constexpr int32_t kPointerPanZoomDeviceId = 1;
 
+// Keeps track of pointer states in relation to the window.
+std::unordered_map<int32_t, std::unique_ptr<PointerState>> pointer_states_;
```

Comment:
This should be inside a class and use GHashTable (it gets too confusing mixing C++ and GObject).